src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+                                Use (batch, total_sequence_length) for shapes of size 2");
+            }
+            inferred_out.index_pad = mask_pad::raw;
+        }
+        else if(mask_index_lens.size() == 3)
+        { // Similar to case 2 but with sequence length in dim 1
+            if(mask_index_lens.at(0) != inferred_out.batch_size or
+               mask_index_lens.at(1) != inferred_out.sequence_length or
+               mask_index_lens.at(2) != inferred_out.total_sequence_length)
+            {
+                MIGRAPHX_THROW("Attention: Invalid Mask_Index shape\n \
+                                Use (batch, sequence_length, total_sequence_length) for shapes of size 3");
+            }
+            inferred_out.index_pad = mask_pad::raw;
+            MIGRAPHX_THROW("Attention: Mask_index 3D masking not supported");
+        }
+        else if(mask_index_lens.size() == 4)
+        { // Oddball case and can be used to infer max_sequence_length_parameter
+            if(mask_index_lens.at(0) != inferred_out.batch_size or mask_index_lens.at(1) != 1 or
+               mask_index_lens.at(2) != mask_index_lens.at(3))
+            {
+                MIGRAPHX_THROW("Attention: Invalid Mask_Index shape\n  \
+                                Use (batch, 1, max_sequence_length, max_sequence_length) for shapes of size 4");
+            }
+            inferred_out.max_sequence_length = mask_index_lens.at(2);
+            inferred_out.index_pad           = mask_pad::raw;
+            MIGRAPHX_THROW("Attention: Mask_index 4D Megatron masking not supported");
+        }
+        else
+        {
+            MIGRAPHX_THROW(
+                "Attention: Mask_index Require shape of size either 1, 2, 3, 4 dimensions");
+        }
+    }
+
+    static void handle_mask_index(const std::vector<instruction_ref>& args,
+                                  struct attention_attr& /*attr_out*/,
+                                  struct attention_inferred& inferred_out,
+                                  std::vector<instruction_ref>& output_arg_vec)
+    {
+        instruction_ref mask_index;
+        if(check_and_return_arg(args, 3, mask_index))
+        {
+            auto mask_index_shape       = mask_index->get_shape();
+            const auto& mask_index_lens = mask_index_shape.lens();
+
+            if(mask_index_shape.type() != migraphx::shape::int32_type)
+            {
+                MIGRAPHX_THROW("Attention: Mask_Index type must be int32 type");
+            }
+            check_mask_index_shapes(mask_index_lens, inferred_out);
+            inferred_out.has_attn_mask = true;
+            output_arg_vec.push_back(mask_index);
+        }
+    }
+
+    static void handle_past(const std::vector<instruction_ref>& args,
+                            const struct attention_attr& attr_out,
+                            struct attention_inferred& inferred_out,
+                            std::vector<instruction_ref>& output_arg_vec)
```

Comment:
Return `std::optional<instruction_ref>` instead of using an output parameter. The vector can be appended in the caller.

Reply:
done